### PR TITLE
Cart stock UX

### LINKED ISSUE #5239: Stuck on cart, if item becomes out of stock during shopping journey
## Description

After a customer visits the shopfront and adds items to cart, it can happen that the stock is reduced, making those items unavailable, before the customer checks out. If this happens, and the customer attempts to check out, he is redirected to the cart page, where the red banner displays with "An item in your cart has become unavailable." Also, in front of the respective item, a red text indicates "Insufficient stock available, only xxx remaining" - pic below.

Now, at this point, the user is "stuck" in the /cart page. Even if the number of items is reduced (either introducing a number or using the small down-arrow) till the "insufficient stock" warning disappears, it is not possible, as the buttons below in [brackets], only leave you on the current cart page:
- to proceed to checkout [ Checkout > ] 
- go back to the shopfront. [ < Continue Shopping ]
- Going "back" in the browser does not work either.

## Expected Behavior

It should be possible to leave the cart page, either by:
- correcting the number of items and pressing  [ Checkout > ]
- go back to the shopfront. [ < Continue Shopping ], for example, to choose from another product

## Actual Behaviour
Customer gets stuck on cart page.

## Steps to Reproduce

C - Customer
A - Admin

C - add X items to the cart.
C - Proceed to Checkout page
A - In Products, reduce the amount of available items, to a value below X
C - Press [Place Order Now]
-> boom!

## Workaround

Here, the only way out, is pressing the trash-bin and removing the items from the cart.

## Severity

bug-s3: a feature is broken but there is a workaround
(however, perhaps this affects few users? S3 and a half? :-)

## Your Environment

* Version used: v2.9.2
* Browser name and version: Firefox 75
* Operating System and version (desktop or mobile): Ubuntu 18.04

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Closes #5239 

Updates various bits of UX to make it clearer to the user where they need to adjust their selected quantities when an item has reduced in stock.

#### What should we test?

UX on cart page when an item's stock has been reduced below the currently selected level.

#### Release notes

Improved cart page UX for carts with insufficient stock

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
- Quantity box highlighted (red border) in the standard warning red colour
- Red border removed when stock is okay
- Update button is highlighted when: any changes have been made to the form **and** the form has no validation errors
- "Continue Shopping" and "Checkout" buttons disabled unless the cart has been updated and is usable

Cart-Stock-UX

**Comment by luisramos0** [comment]
Hey @Matt-Yorkley can you update the messages as well with a cta of some kind: "Please adjust the quantity and press Update" This can be on the flash and maybe also on the line item message.

File: app/assets/javascripts/darkswarm/controllers/cart_form_controller.js.coffee [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Darkswarm.controller "CartFormCtrl", ($scope) ->
+
```
Comment (luisramos0) [inline app/assets/javascripts/darkswarm/controllers/cart_form_controller.js.coffee:2]: ah, this bootstraps angular basics on the form? I didnt know this :ok_hand:

File: app/assets/javascripts/darkswarm/directives/validate_stock_quantity.js.coffee [hunk 2]
```diff
@@ -0,0 +1,17 @@
+Darkswarm.directive "validateStockQuantity", ->
+  restrict: 'A'
+  require: "ngModel"
+
+  link: (scope, element, attr, ngModel) ->
+    ngModel.$parsers.push (selectedQuantity) ->
+      if parseInt(selectedQuantity) > scope.available_quantity()
+        ngModel.$setValidity('stock', false);
+      else
+        ngModel.$setValidity('stock', true);
```
Comment (luisramos0) [inline app/assets/javascripts/darkswarm/directives/validate_stock_quantity.js.coffee:10]:
```javascript
      valid? = parseInt(selectedQuantity) > scope.available_quantity()
      ngModel.$setValidity('stock', valid?);
```

Comment (luisramos0) [reply app/assets/javascripts/darkswarm/directives/validate_stock_quantity.js.coffee:10]: re parseInt - what happens if user types a letter in the quantity field?

File: app/assets/javascripts/darkswarm/directives/validate_stock_quantity.js.coffee [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+      selectedQuantity
+
+    scope.available_quantity = ->
+      on_hand = parseInt(attr.ofnOnHand)
+      finalized_quantity = parseInt(attr.finalizedquantity) || 0 # finalizedquantity is optional
+      on_hand + finalized_quantity
```
Comment (luisramos0) [inline app/assets/javascripts/darkswarm/directives/validate_stock_quantity.js.coffee:17]: I think this is complicated code already in ofnOnHand, I dont think we can copy paste this type of logic. It needs to be extracted to a service to be re-used here.

File: app/assets/javascripts/darkswarm/directives/disable_dynamically.js.coffee [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Allows disabling of link buttons via disabled attribute.
+# This is normally ignored, ie the link appears disabled but is still clickable.
+
+Darkswarm.directive "disableDynamically", ->
+  restrict: 'A'
+
+  link: (scope, element, attrs) ->
+    element.on 'click', (e) ->
+      console.log(attrs)
```
Comment (luisramos0) [inline app/assets/javascripts/darkswarm/directives/disable_dynamically.js.coffee:9]: this needs to be deleted right?

File: app/assets/javascripts/darkswarm/directives/disable_dynamically.js.coffee [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# Allows disabling of link buttons via disabled attribute.
+# This is normally ignored, ie the link appears disabled but is still clickable.
```
Comment (luisramos0) [inline app/assets/javascripts/darkswarm/directives/disable_dynamically.js.coffee:2]: are you saying "ng-disabled"" doesnt work in these cases? do you know why? It looks like this directive shouldnt be needed.

File: app/assets/javascripts/darkswarm/controllers/cart_form_controller.js.coffee (hunk 1 above)
Comment (Matt-Yorkley) [reply app/assets/javascripts/darkswarm/controllers/cart_form_controller.js.coffee:2]: I think if there's no controller on an element, there's no `$scope` to play with. Initially I thought I would need to add some more code to this controller over the course of the PR, but in the end there was nothing else needed.

File: app/assets/javascripts/darkswarm/directives/disable_dynamically.js.coffee (hunk 5 above)
Comment (Matt-Yorkley) [reply app/assets/javascripts/darkswarm/directives/disable_dynamically.js.coffee:2]: It does not work! It appears to be disabled, but is still clickable. I tried hacking the `href` property instead with a conditional so the link would not function when the conditions are not met, but that didn't work either. :see_no_evil: 

SO discussion here: https://stackoverflow.com/questions/23425254/enable-disable-anchor-tags-using-angularjs

File: app/assets/javascripts/darkswarm/directives/validate_stock_quantity.js.coffee (hunk 3 above)
Comment (Matt-Yorkley) [reply app/assets/javascripts/darkswarm/directives/validate_stock_quantity.js.coffee]: I've made this more explicit :+1:

Comment (Matt-Yorkley) [reply app/assets/javascripts/darkswarm/directives/validate_stock_quantity.js.coffee]: Done.

File: app/assets/javascripts/darkswarm/directives/validate_stock_quantity.js.coffee (hunk 2 above)
Comment (Matt-Yorkley) [reply app/assets/javascripts/darkswarm/directives/validate_stock_quantity.js.coffee]: I've made this more explicit :+1:

File: app/assets/javascripts/darkswarm/services/stock_quantity.js.coffee [hunk 6]
```diff
@@ -0,0 +1,7 @@
+Darkswarm.factory "StockQuantity", ->
```
Comment (luisramos0) [inline app/assets/javascripts/darkswarm/services/stock_quantity.js.coffee:1]: nice, thank you!!!

**Comment by filipefurtad0** [comment]
This looks awesome @Matt-Yorkley,

After this PR, upon stock reduction during checkout:
- the warning in the red banner "An item in your cart has become unavailable. Please update the selected quantities." is shown
- the "Qty" field gets highlighted in red
- after reducing the value to an allowed number (in accordance with the modal) the update button get's fully red - can't miss it.
- while in this state, the Continue Shopping and Checkout buttons are greyed-out

This is good to go :tada:

**Pull request merged**